### PR TITLE
[FLINK-27263][table] Rename the metadata column to the user specified…

### PR DESCRIPTION
**Pull request by fsk119**
… name in DDL

## What is the purpose of the change

*Currently the source output the metadata column with the `"$metadata"+ metadata key` as the column name. It's better that keep the name align with the user name specified in the DDL,e.g. use `event_time` as the name if the user specifies `event_time TIMESTAMP(3) FROM 'timestamp'`.*

## Brief change log

  - *Make call in the projection to do the rename*

## Verifying this change

This change is already covered by existing tests in the `PushProjectIntoTableSourceScanRuleTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Review by wuchong — COMMENTED** [review]
LGTM.

Would be better @twalthr have a look.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/MergeTableLikeUtilTest.java [hunk 1]
```diff
@@ -431,6 +433,100 @@ public void mergeOverwritingMetadataColumnsDuplicate() {
...
+        List<SqlNode> derivedColumns =
+                Arrays.asList(
+                        metadataColumn("one", DataTypes.BOOLEAN(), true),
+                        metadataColumn("two", DataTypes.STRING(), "one", false));
+
+        Map<FeatureOption, MergingStrategy> mergingStrategies = getDefaultMergingStrategies();
+        mergingStrategies.put(FeatureOption.METADATA, MergingStrategy.INCLUDING);
+
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/MergeTableLikeUtilTest.java:449]: How about we add this validation logic to `DefaultSchemaResolver`? This way we avoid checking duplicate metadata keys at three locations and fail as early as possible.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @fsk119. The changes look very good and better than the previous solution. I had one last comment to simplify the implementation a bit. Should be good to merge afterwards.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java [hunk 2]
```diff
@@ -168,26 +165,40 @@ public static void prepareDynamicSource(
...
-        final Set<String> requiredMetadataKeys =
-                metadataColumns.stream()
-                        .map(c -> c.getMetadataKey().orElse(c.getName()))
-                        .collect(Collectors.toSet());
+        Map<String, MetadataColumn> metadataKeysToMetadataColumns = new HashMap<>();
+
+        for (MetadataColumn column : metadataColumns) {
+            String metadataKey = column.getMetadataKey().orElse(column.getName());
+            if (metadataKeysToMetadataColumns.containsKey(metadataKey)) {
+                throw new ValidationException(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSourceUtils.java]: We don't need this check here. `DefaultSchemaResolver` should be the only location for validation.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala [hunk 3]
```diff
@@ -215,6 +215,87 @@ class TableEnvironmentTest {
...
+        |  b INT METADATA FROM 'a'
+        |)
+        |WITH (
+        |  'connector' = 'COLLECTION'
+        |) LIKE source (
+        |  EXCLUDING METADATA
+        |)
+        |""".stripMargin
+    )
+    assertEquals(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala:288]: not sure if testing this is still necessary. this should be covered by tests already?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala [hunk 4]
```diff
@@ -215,6 +215,87 @@ class TableEnvironmentTest {
       TableTestUtil.replaceNodeIdInOperator(TableTestUtil.replaceStreamNodeId(actual)))
   }
 
+  @Test
+  def testCreateTableWithMultipleColumnsFromSameMetadataKey(): Unit = {
```
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala]: Move to the TableScanTest.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala (hunk 3 above)
Comment (fsk119) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala]: Remove the test.

**Review by twalthr — APPROVED** [review]
There is one minor mistake in the JavaDocs, but I will fix this during merge. LGTM otherwise.

**Comment by fsk119** [comment]
@flinkbot  run azure

Thanks for your review and help =-=.

**Comment by twalthr** [comment]
Fixed in 1.15 via 5e9ccf1cdd4328885ac50fad4d32fc93a62290fd.

**Pull request closed without merging**